evm/src/cpu/kernel/asm/curve_add.asm

```diff
@@ -0,0 +1,383 @@
+    // stack: ret_zero, (x1,y1)==(0,0), x1, y1, retdest
+    JUMPI
+    // stack: x1, y1, retdest
+    SWAP1
+    // stack: y1, x1, retdest
+    SWAP2
+    // stack: retdest, x1, y1
+    JUMP
+
+// Assumption: (x1,y1) == (0,0) and (x0,y0) != (0,0)
+ec_add_snd_zero:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest
+    SWAP2
+    // stack: x1, y0, x0, y1, retdest
+    POP
+    // stack: y0, x0, y1, retdest
+    SWAP2
+    // stack: y1, x0, y0, retdest
+    POP
+    // stack: x0, y0, retdest
+    SWAP1
+    // stack: y0, x0, retdest
+    SWAP2
+    // stack: retdest, x0, y0
+    JUMP
+
+ret_zero:
+    JUMPDEST
+    // stack: x, y, retdest
+    POP
+    // stack: y, retdest
+    POP
+    // stack: retdest
+    PUSH 0
+    // stack: 0, retdest
+    PUSH 0
+    // stack: 0, 0, retdest
+    SWAP2
+    // stack:  retdest, 0, 0
+    JUMP
+
+ec_add_valid_points_with_lambda:
+    JUMPDEST
+    // stack: lambda, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x0, lambda, x0, y0, x1, y1, retdest
+    DUP5
+    // stack: x1, x0, lambda, x0, y0, x1, y1, retdest
+    %bn_base
+    // stack: N, x1, x0, lambda, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: lambda, N, x1, x0, lambda, x0, y0, x1, y1, retdest
+    DUP1
+    // stack: lambda, lambda, N, x1, x0, lambda, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: lambda^2, x1, x0, lambda, x0, y0, x1, y1, retdest
+    %submod
+    // stack: lambda^2 - x1, x0, lambda, x0, y0, x1, y1, retdest
+    %submod
+    // stack: x2, lambda, x0, y0, x1, y1, retdest
+    %bn_base
+    // stack: N, x2, lambda, x0, y0, x1, y1, retdest
+    DUP2
+    // stack: x2, N, x2, lambda, x0, y0, x1, y1, retdest
+    DUP7
+    // stack: x1, x2, N, x2, lambda, x0, y0, x1, y1, retdest
+    %submod
+    // stack: x1 - x2, N, x2, lambda, x0, y0, x1, y1, retdest
+    DUP4
+    // stack: lambda, x1 - x2, N, x2, lambda, x0, y0, x1, y1, retdest
+    MULMOD
+    // stack: lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    DUP7
+    // stack: y1, lambda * (x1 - x2), x2, lambda, x0, y0, x1, y1, retdest
+    SWAP1
+    // stack: lambda * (x1 - x2), y1, x2, lambda, x0, y0, x1, y1, retdest
+    %submod
+    // stack: y2, x2, lambda, x0, y0, x1, y1, retdest
+    SWAP5
+    // stack: x1, x2, lambda, x0, y0, y2, y1, retdest
+    POP
+    // stack: x2, lambda, x0, y0, y2, y1, retdest
+    SWAP5
+    // stack: y1, lambda, x0, y0, y2, x2, retdest
+    POP
+    // stack: lambda, x0, y0, y2, x2, retdest
+    POP
+    // stack: x0, y0, y2, x2, retdest
+    POP
+    // stack: y0, y2, x2, retdest
+    POP
+    // stack: y2, x2, retdest
+    SWAP2
+    // stack: retdest, x2, y2
+    JUMP
+
+// Assumption: (x0,y0) and (x1,y1) are valid points and x0 == x1
+ec_add_equal_first_coord:
+    JUMPDEST
+    // stack: x0, y0, x1, y1, retdest with x0 == x1
+    %bn_base
+    // stack: N, x0, y0, x1, y1, retdest
+    DUP3
+    // stack: y0, N, x0, y0, x1, y1, retdest
+    DUP6
+    // stack: y1, y0, N, x0, y0, x1, y1, retdest
+    ADDMOD
```

Comment:
nit: `EQ` seems a bit more intuitive, and should be a tad cheaper since it's done directly on the CPU